pkg/provisioner/ironic/ironic.go

```diff
@@ -124,6 +160,19 @@ func newProvisioner(host *metal3v1alpha1.BareMetalHost, bmcCreds bmc.Credentials
 	if err != nil {
 		return nil, err
 	}
+	if ironicUseMutualTLS {
```

Comment:
It feels a little odd to instantiate a NoAuth client and then modify its HTTPClient. Is that the way it's supposed to be used?

I see that gophercloud has a different client implementation for basic auth in https://github.com/gophercloud/gophercloud/blob/master/openstack/baremetal/httpbasic/requests.go. Would it make sense to add this new client type to gophercloud? Or do we expect this use pattern to be unique to metal3?

If this is something we only want in metal3, maybe we should move the ironic client creation code into its own function that can be invoked here to separate the details of setting up those clients from the provisioner.

Reply:
I would be reluctant to modify gophercloud considering that this is not something supported by Ironic, but an Apache-based setup that we are doing. I'd say it's specific to Metal3. The whole clients creation and adaptation should be indeed moved to its own function. We could implement the exact same thing in BMO, creating our own type of clients like https://github.com/gophercloud/gophercloud/blob/master/openstack/baremetal/httpbasic/requests.go

Reply:
the example followed is in the FAQ : https://github.com/gophercloud/gophercloud/blob/master/docs/FAQ.md#implementing-default-logging-and-re-authentication-attempts

Reply:
Keystone supports auth using client certificates, and python-openstackclient supports using client certificates. So I don't actually think it would be unreasonable to add this in gophercloud. (Probably no need to block on that though.)